File: hw2/python/hw2.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import numpy as np
from math import sqrt


def main():
    # Define coefficient matrix
    A = np.matrix([
        [1, 0, 0, 1, 0, -1, 0],
        [0, -1, 0, 1, 1, 0, 0],
        [0, 1, 0, 0, 0, 0, 1],
        [1, 1, 1, 0, 0, 0, 0]])

    # Define sigma0 and weight matrix
    s0 = 3.0
    W = np.matrix([
        [s0**2/.5**2, 0, 0, 0, 0, 0, 0],
        [0, s0**2/.1**2, 0, 0, 0, 0, 0],
        [0, 0, s0**2/.1**2, 0, 0, 0, 0],
        [0, 0, 0, s0**2/.3**2, 0, 0, 0],
        [0, 0, 0, 0, s0**2/.2**2, 0, 0],
        [0, 0, 0, 0, 0, s0**2/.3**2, 0],
        [0, 0, 0, 0, 0, 0, s0**2/.1**2]])

    # Define f matrix
    f = np.matrix([
        [.3],
        [0],
        [0],
        [.1]])

    # Define vector of obserbation
    l = np.matrix([
        [1.2],
        [2.4],
        [-3.7],
        [-.4],
        [2.8],
        [1.1],
        [-2.4]])

    # Compute Normal matrix
    N = A * W.I * A.T

    # Compute K matrix
    K = N.I * f

    # Compute residual vector
    V = W.I * A.T * K
    print "Residual : "
    print "V = \t", str(V.round(4)).replace("\n", "\n\t")
    print

    # Compute corrected observation
    L = l + V
    print "Correct observation : "
    print "L = \t", str(L.round(4)).replace("\n", "\n\t")
    print

    # Compute error of unit weight
    sigma0 = sqrt((V.T * W * V) / A.shape[0])
    print "Error of unit weight : %.4f" % sigma0

if __name__ == '__main__':
    main()

File: hw2/matlab/hw2.m
clear;
clc;

% Define coefficient matrix
A = [1 0 0 1 0 -1 0;
     0 -1 0 1 1 0 0;
     0 1 0 0 0 0 1;
     1 1 1 0 0 0 0];

% Define sigma0 and weight matrix
s0 = 3.0;
w = [.5 .1 .1 .3 .2 .3 .1];
W = diag(s0./w)^2;

% Define f matrix
f = [.3 0 0 .1]';

% Define vector of obserbation
l = [1.2 2.4 -3.7 -.4 2.8 1.1 -2.4]';

% Compute Normal matrix
N = A * inv(W) * A';

% Compute K matrix
K = inv(N) * f;

% Compute residual vector
V = inv(W) * A' * K;

% Print results
disp('Residual : ');
disp(V);

% Compute corrected observation
L = l + V;
disp('Correct observation : ');
disp(L);

% Compute error of unit weight
[row, col] = size(A);
sigma0 = sqrt((V'* W * V) / row);
fprintf('Error of unit weight : %.4f\n', sigma0);
